test/tests/pbs_execjob_hook_pp350.py

```diff
@@ -0,0 +1,98 @@
+# coding: utf-8
+
+# Copyright (C) 1994-2016 Altair Engineering, Inc.
+# For more information, contact Altair at www.altair.com.
+#
+# This file is part of the PBS Professional ("PBS Pro") software.
+#
+# Open Source License Information:
+#
+# PBS Pro is free software. You can redistribute it and/or modify it
+# under the terms of the GNU Affero General Public License as published
+# by the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# PBS Pro is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public
+# License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+# Commercial License Information:
+#
+# The PBS Pro software is licensed under the terms of the GNU Affero
+# General Public License agreement ("AGPL"), except where a separate
+# commercial license agreement for PBS Pro version 14 or later has been
+# executed in writing with Altair.
+#
+# Altair’s dual-license business model allows companies, individuals,
+# and organizations to create proprietary derivative works of PBS Pro
+# and distribute them - whether embedded or bundled with other software
+# - under a commercial license agreement.
+#
+# Use of Altair’s trademarks, including but not limited to "PBS™",
+# "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to
+# Altair's trademark licensing policies.
+
+from ptl.utils.pbs_testsuite import *
+
+
+class ExecJobHookTest(PBSTestSuite):
+
+    hooks = {
+        "execjob_hook1":
+        """
+import pbs
+pbs.logmsg(pbs.LOG_DEBUG, "executing execjob_launch 1" )
+e=pbs.event()
+e.progname = "/usr/bin/echo"
+e.argv=[]
+e.argv.append("launch 1")
+        """,
+        "execjob_hook2":
+        """
+import pbs
+pbs.logmsg(pbs.LOG_DEBUG, "executing execjob_launch 2" )
+e = pbs.event()
+if (e.progname != "/usr/bin/echo"):
+    pbs.logmsg(pbs.LOG_DEBUG,
+        "Modified progname value did not propagated from lauch1 hook")
```

Comment:
I agree, the indentation is off, particularly line 3013.
- We don't need to free the hook input values, because the way the execjob_launch hook processing is done takes care of freeing any current and previously set values for hook_input and hook_output.

The way hook_input and hook_output works is as follows.
a. hook_input values are fed to run_hook() call which launches the pbs_python process. The hook_output values are set in get_hook_results().
b. The initial values set for hook_input->progname, argv, envp (seen if num_run == 0) are automatically freed up in start_exec.c:finish_exec() after mom_process_hooks() has been executed in a child mom process servicing the job:
      mom_hook_input_init(&hook_input);
      hook_input.progname = the_progname;
      hook_input.argv = the_argv;
      hook_input.env = the_env;
...

        switch (mom_process_hooks(HOOK_EVENT_EXECJOB_LAUNCH,
                        PBS_MOM_SERVICE_NAME,
                        mom_host, &hook_input, &hook_output,
                                hook_msg, sizeof(hook_msg), 0))
        {

...

                execve(the_progname, the_argv, the_env);
                free(progname);
                free_str_array(the_argv);
                the_argv = NULL;
                free_str_array(the_env);
                the_env = NULL;
             <-- Note that there's no freeing of the_progname here as this is set 
                     to /etc/passwd shell for the user or the PBS Job's Shell_Path entry. 
                      Management of memory for that is done somewhere else.

c. The last known hook_output values are also automatically freed up after mom_process_hooks() is called responding to an EXECJOB_LAUNCH event under start_exec.c:finish_exec():

        mom_hook_output_init(&hook_output);
        hook_output.progname = &progname;
        hook_output.argv = &argv_list;
        hook_output.env = &env_str;
...
       
     switch (mom_process_hooks(HOOK_EVENT_EXECJOB_LAUNCH,
                        PBS_MOM_SERVICE_NAME,
                        mom_host, &hook_input, &hook_output,
                                hook_msg, sizeof(hook_msg), 0))
        {

         execve(the_progname, the_argv, the_env);
         free(progname);
         free_attrlist(&argv_list);
        free(env_str);

d. Now going back to mom_process_hooks(), we see the non-initial values of hook_input are changed to interim hook_output values. These interim values (except the last values which are freed in start_exec.c:finish_exec()) are automatically freed by the next call to get_hook_results():

...
                        } else if (strcmp(name_str, PY_EVENT_PARAM_PROGNAME) == 0) {
                                char    **prog;
                                if (hook_output != NULL) {
                                        /* need to free up here previous value */
                                        /* in case of multiple hooks! */
                                        prog = hook_output->progname;
                                        if (*prog != NULL) {
                                                free(*prog);
                                        }
                                        *prog = strdup(data_value);
                                }
                        } else if (strcmp(name_str, PY_EVENT_PARAM_ARGLIST) == 0) {
                                pbs_list_head *ar_list;

                                arg_list_entries++;
                                if (hook_output != NULL) {
                                        ar_list = hook_output->argv;
                                        /* free previous values at start of new list */
                                        if (arg_list_entries == 1) {
                                                free_attrlist(ar_list);
                                        }
                                        add_to_svrattrl_list(ar_list, name_str, resc_str,
                                                                        data_value, 0, NULL);
                                }
                        } else if (strcmp(name_str, PY_EVENT_PARAM_ENV) == 0) {
                                char    **env;
                                if (hook_output != NULL) {
                                        env = hook_output->env;
                                        if (*env != NULL) {
                                                free(*env);
                                        }
                                        *env = strdup(data_value);
                                }
                        }

 
e. Even if we missed freeing something, the mom_process_hooks() call is being done by a child mom launched in behalf of a job. If ever there was a memory leak, it's only contained in that child mom, which will go away when the job finishes executing or aborts.